compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1231,25 +1231,38 @@ pub fn linker_and_flavor(sess: &Session) -> (PathBuf, LinkerFlavor) {
                     sess.emit_fatal(errors::LinkerFileStem);
                 });
 
+                // Some linker flavors can have an optional version postfix. GCC can have an
+                // optional target prefix.
                 let flavor = if stem == "emcc" {
                     LinkerFlavor::EmCc
                 } else if stem == "gcc"
+                    || stem.starts_with("gcc-")
                     || stem.ends_with("-gcc")
+                    || stem.contains("-gcc-")
                     || stem == "clang"
-                    || stem.ends_with("-clang")
+                    || stem.starts_with("clang-")
                 {
                     LinkerFlavor::from_cli(LinkerFlavorCli::Gcc, &sess.target)
-                } else if stem == "wasm-ld" || stem.ends_with("-wasm-ld") {
+                } else if stem == "wasm-ld"
+                    || stem.ends_with("-wasm-ld")
+                    || stem.starts_with("wasm-ld-")
+                {
                     LinkerFlavor::WasmLld(Cc::No)
                 } else if stem == "ld" || stem.ends_with("-ld") {
                     LinkerFlavor::from_cli(LinkerFlavorCli::Ld, &sess.target)
-                } else if stem == "ld.lld" {
+                } else if stem == "ld.lld" || stem.starts_with("ld.lld-") {
                     LinkerFlavor::Gnu(Cc::No, Lld::Yes)
                 } else if stem == "link" {
                     LinkerFlavor::Msvc(Lld::No)
-                } else if stem == "lld-link" {
+                } else if stem == "lld-link" || stem.starts_with("lld-link-") {
                     LinkerFlavor::Msvc(Lld::Yes)
-                } else if stem == "lld" || stem == "rust-lld" {
+                } else if stem == "lld"
+                    || stem
+                        .strip_prefix("lld-")
+                        .filter(|x| x.chars().all(char::is_numeric))
+                        .is_some()
+                    || stem == "rust-lld"
+                {
```

Comment:
Did you actually encounter the versioned variants of lld in practice?

Reply:
FWIW, lld's own flavor detection logic can be found in https://github.com/llvm/llvm-project/blob/main/lld/tools/lld/lld.cpp#L110-L122

Reply:
> Did you actually encounter the versioned variants of lld in practice?

Yes, e.g. on ubuntu lld with the version postfix is part of the [llvm-toolchain-xx](https://packages.ubuntu.com/source/jammy/llvm-toolchain-12) package. Since different people on our team are using different versions of ubuntu or other distributions, we generally prefer using tools with an explicit version postfix (in CI) to make it clear which version of a tool is known to be working.